File: leetcode/383. Ransom Note.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char,int> um;
        
        for(int i = 0 ; i < magazine.size() ; i++){
            um[magazine[i]]++;
        }
        
        for(int i = 0 ; i < ransomNote.size() ; i++){
            if(um.count(ransomNote[i]) <= 0){
                return false;
            } else {
                if(um[ransomNote[i]] <= 0) return false;
                um[ransomNote[i]]--;
            }      
        }
        return true;
    }
};

File: online-training-platform/leetcode/383. Ransom Note.kt
class Solution {

    // Time O(m + n)
    // Space O(26) = O(1)
    fun canConstruct(ransomNote: String, magazine: String): Boolean {
        val ransomNoteCount = ransomNote.groupingBy{it}.eachCount()
        val magazineCount = magazine.groupingBy{it}.eachCount()

        var result = true
        ransomNoteCount.forEach{ letter, count ->
            if(magazine.contains(letter).not()){
                result = false
                return@forEach 
            }

            if(magazineCount.getOrDefault(letter, 0) < count){
                result = false
                return@forEach
            }
        }

        return result
    }
}